transaction_decoder/transaction_decoder.py

```diff
@@ -0,0 +1,287 @@
+
+        result = TransactionMetadata()
+        if not result.transfers:
+            result.transfers = []
+
+        index = 2
+        for _ in range(transfer_count):
+            identifier = self.hex_to_string(args[index])
+            index += 1
+            nonce = args[index]
+            index += 1
+            value = self.hex_to_big_int(args[index])
+            index += 1
+
+            if nonce:
+                tx_metadata = TransactionMetadataTransfer()
+
+                tx_metadata.value = value
+                tx_metadata.properties = TokenTransferProperties()
+                tx_metadata.properties.collection = identifier
+                tx_metadata.properties.identifier = f"{identifier}-{nonce}"
+
+                result.transfers.append(tx_metadata)
+            else:
+                tx_metadata = TransactionMetadataTransfer()
+
+                tx_metadata.value = value
+                tx_metadata.properties = TokenTransferProperties()
+                tx_metadata.properties.collection = identifier
+
+                result.transfers.append(tx_metadata)
+
+        result.sender = metadata.sender
+        result.receiver = receiver
+
+        if len(args) > index:
+            result.function_name = self.hex_to_string(args[index])
+            index += 1
+            result.function_args = args[index:]
+            index += 1
+
+        return result
+
+    def getEsdtTransactionMetadata(
+        self, metadata: TransactionMetadata
+    ) -> Optional[TransactionMetadata]:
+        if metadata.function_name != "ESDTTransfer":
+            return None
+
+        args = metadata.function_args
+        if not args:
+            return None
+
+        if len(args) < 2:
+            return None
+
+        token_identifier = self.hex_to_string(args[0])
+        value = self.hex_to_big_int(args[1])
+
+        result = TransactionMetadata()
+
+        result.sender = metadata.sender
+        result.receiver = metadata.receiver
+        result.transfers = []
+
+        if len(args) > 2:
+            result.function_name = self.hex_to_string(args[2])
+            result.function_args = args[3:]
+
+        tx_metadata = TransactionMetadataTransfer()
+        tx_metadata.value = value
+
+        tx_metadata.properties = TokenTransferProperties()
+        tx_metadata.properties.collection = token_identifier
+        tx_metadata.properties.identifier = token_identifier
+
+        result.transfers.append(tx_metadata)
+        result.value = value
+
+        return result
+
+    def is_address_valid(self, address: str) -> bool:
+        return len(binascii.unhexlify(address)) == 32
+
+    def is_smart_contract_argument(self, arg: str) -> bool:
+        if not self.is_hex(arg):
+            return False
+        if len(arg) % 2 != 0:
+            return False
+        return True
+
+    def is_hex(self, value: str) -> bool:
+        return not bool(re.search("[^a-f0-9]", value, re.IGNORECASE))
```

Comment:
not sure about the necessity of speed, but iterating over each character instead of a regex is much faster